include/ngp_utils/NgpFieldOps.h

```diff
@@ -18,163 +21,309 @@
-  void set(const int n, const int ic, const DoubleType& value) const
+  KOKKOS_INLINE_FUNCTION
+  const Ops& operator()(const int n, const int ic = 0) const
   {
-#ifndef KOKKOS_ENABLE_CUDA
-    for (int is=0; is < edata_.numSimdElems; ++is) {
-      ngpField_.get(ngpMesh_, edata_.elemInfo[is].entityNodes[n], ic) =
-        stk::simd::get_data(value, is);
-    }
-#else
-    ngpField_.get(ngpMesh_, edata_.elemInfo[0].entityNodes[n], ic) =
-      stk::simd::get_data(value, 0);
-#endif
+    ops_.ni = n;
+    ops_.ic = ic;
+    return ops_;
   }
 
-  /** Add to the value of a NODE_RANK field from within an element loop
-   *
-   *  This method adds to the i-th component of a nodal quantity belonging to
-   *  the n-th node connected to an element from within an element loop.
-   *
-   *  @param n The node index into the element connectivity array
-   *  @param ic The component index for the field data array
-   *  @param value The SIMD data to to be atomically added
+  //! NGP Mesh instance
+  const Mesh& ngpMesh_;
+
+  //! NGP element field to be updated
+  const Field& ngpField_;
+
+  //! Connectivity data for SIMD group
+  const ElemSimdData<Mesh>& edata_;
+
+  mutable Ops ops_;
+};
+
+/** Update an NGP field registered to ELEM_RANk from within SIMD-ized loop
+ */
+template<typename Mesh, typename Field>
+struct ElemFieldOp
+{
+  static_assert(std::is_floating_point<typename Field::value_type>::value,
+                "NGP field must have a floating type");
+
+  KOKKOS_INLINE_FUNCTION
+  ElemFieldOp(
+    const Mesh& ngpMesh,
+    const Field& ngpField,
+    const ElemSimdData<Mesh>& elemData
+  ) : ngpMesh_(ngpMesh),
+      ngpField_(ngpField),
+      edata_(elemData),
+      ops_(*this)
+  {}
+
+  KOKKOS_FUNCTION ~ElemFieldOp() = default;
+
+  /** Implementation of the operators
    */
-  KOKKOS_FORCEINLINE_FUNCTION
-  void add(const int n, const int ic, const DoubleType& value) const
+  struct Ops
   {
-#ifndef KOKKOS_ENABLE_CUDA
-    for (int is=0; is < edata_.numSimdElems; ++is) {
-      Kokkos::atomic_add(
-        &ngpField_.get(ngpMesh_, edata_.elemInfo[is].entityNodes[n], ic),
-        stk::simd::get_data(value, is));
+    KOKKOS_INLINE_FUNCTION
+    Ops(ElemFieldOp<Mesh, Field>& obj) : obj_(obj)
+    {}
+
+    KOKKOS_INLINE_FUNCTION
+    void operator= (const DoubleType& val) const
+    {
+      const auto& fld = obj_.ngpField_;
+      const auto* einfo = obj_.edata_.elemInfo;
+#ifdef STK_SIMD_NONE
+      fld.get(einfo[0].meshIdx, ic) = stk::simd::get_data(val, 0);
+#else
+      for (int is=0; is < obj_.edata_.numSimdElems; ++is) {
+        fld.get(einfo[is].meshIdx, ic) = stk::simd::get_data(val, is);
+      }
+#endif
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    void operator+= (const DoubleType& val) const
+    {
+      const auto& fld = obj_.ngpField_;
+      const auto* einfo = obj_.edata_.elemInfo;
+#ifdef STK_SIMD_NONE
+      Kokkos::atomic_add(fld.get(einfo[0].meshIdx, ic), stk::simd::get_data(val, 0));
```

Comment:
It seems like we shouldn't need atomic-add for element fields. We definitely do for nodal fields since nodes can be shared by different elements which are being worked on by different threads at the same time. But for element fields our current parallelism doesn't have two threads working on the same element at the same time.

Reply:
@alanw0 That is a great point. I will update the code to reflect our current looping pattern/strategy.